### PR TITLE
add a batch norm inference kernel.

### LINKED ISSUE #929: batch norm 预测时指定 use_global_stats=True，出现 CUDNN_STATUS_NOT_SUPPORTED 错误
文本分类任务，想使用 cnn + batch norm，配置的片段如下：

```python
Layer(name=name + 'context1',
          type="mixed",
          bias=False,
          inputs=ContextProjection(input_name,
                                   context_start=0,
                                   context_length=window_size,
                                   trainable_padding=False))

    Layer(name=name + 'conv0',
          type="mixed",
          size=size,
          active_type="linear",
          bias=Bias(initial_std=1e-1,
                    initial_mean=0,
                    is_static=static,
                    learning_rate=lr),
          inputs=[FullMatrixProjection(name + "context1",
                                       initial_std=2e-2,
                                       is_static=static,
                                       learning_rate=lr)])

    Layer(name=name + 'batch_norm0',
          type='batch_norm',
          active_type="relu",
          use_global_stats=True,
          bias=Bias(initial_mean=0.1, initial_std=0,
              is_static=static, learning_rate=lr),
          inputs=Input(name + 'conv0',
                       initial_mean=1.0,
                       initial_std=0.0,
                       is_static=static,
                       learning_rate=lr,
                       image=Image(channels=size, img_size=1)), )
```
使用 gpu 训练可以正常训练。
测试时，希望保持 use_global_stats=True，使用训练时存储下来的 mean 和 std 的working average ，
但是报如下错误：

image

### PR DESCRIPTION
**Pull request by qingqing01**
Fix #929

### CODE REVIEW

**Comment by qingqing01** [comment]
可以使用下面代码验证: 

```
#include <cuda.h>
#include <cudnn.h>
#include <iostream>
#include <sstream>
#include <fstream>

#include <stdio.h>

#define TOSTR_(s)   #s
#define TOSTR(s)    TOSTR_(s)
#define CUDNN_VERSION_STR  TOSTR(CUDNN_MAJOR) "." TOSTR (CUDNN_MINOR) "." TOSTR(CUDNN_PATCHLEVEL)

#define FatalError(s) {                                                \
    std::stringstream _where, _message;                                \
    _where << __FILE__ << ':' << __LINE__;                             \
    _message << std::string(s) + "\n" << __FILE__ << ':' << __LINE__;\
    std::cerr << _message.str() << "\nAborting...\n";                  \
    cudaDeviceReset();                                                 \
    exit(EXIT_FAILURE);                                                \
}

#define checkCUDNN(status) {                                           \
    std::stringstream _error;                                          \
    if (status != CUDNN_STATUS_SUCCESS) {                              \
      _error << "CUDNN failure\nError: " << cudnnGetErrorString(status); \
      FatalError(_error.str());                                        \
    }                                                                  \
}

#define checkCUDA(status) {                                      \
    std::stringstream _error;                                          \
    if (status != 0) {                                                 \
      _error << "Cuda failure\nError: " << cudaGetErrorString(status); \
      FatalError(_error.str());                                        \
    }                                                                  \
}



#include <sys/time.h>
#include <unistd.h>


void create(float** h_v, float** d_v, int n) {
  *h_v = (float *)malloc(n * sizeof(float));
  checkCUDA(cudaMalloc(d_v, n * sizeof(float)));
  for(int i = 0; i < n; i++)
    (*h_v)[i] = 1.0f;
  checkCUDA(cudaMemcpy(*d_v, *h_v, n * sizeof(float), cudaMemcpyHostToDevice));
}

int main(int argc, char *argv[]) {   

  int version = (int)cudnnGetVersion();
  printf("cudnnGetVersion() : %d , CUDNN_VERSION from cudnn.h : %d (%s)\n",
      version, CUDNN_VERSION, CUDNN_VERSION_STR);
  cudaSetDevice(0);

  /* input dim */
  int n, c, h, w;
  n = 1025;
  c = 512;
  h = 1;
  w = 1;

  /* Handles */
  cudnnHandle_t cudnnHandle;
  cudnnTensorDescriptor_t ioDesc, bnDesc;
  
  /* Create Handles and Descriptor*/
  checkCUDNN( cudnnCreate(&cudnnHandle));
  checkCUDNN( cudnnCreateTensorDescriptor(&ioDesc));
  checkCUDNN( cudnnCreateTensorDescriptor(&bnDesc));

  /* some constants */
  cudnnDataType_t dataType = CUDNN_DATA_FLOAT;
  cudnnBatchNormMode_t mode = CUDNN_BATCHNORM_SPATIAL;

  /* initilize input and output buffers */
  float* h_input;
  float* d_input;
  float* h_scale;
  float* d_scale;
  float* h_bias;
  float* d_bias;
  float* h_estimated_mean;
  float* d_estimated_mean;
  float* h_estimated_var;
  float* d_estimated_var;

  float* h_output;
  float* d_output;

  create(&h_input, &d_input, n * c * h * w);
  create(&h_output, &d_output, n * c * h * w);
  create(&h_scale, &d_scale, c);
  create(&h_bias, &d_bias, c);
  create(&h_estimated_mean, &d_estimated_mean, c);
  create(&h_estimated_var, &d_estimated_var, c);

  /* initilize handles */
  const int stride_w = 1;
  const int stride_h = w * stride_w;
  const int stride_c = h * stride_h;
  const int stride_n = c * stride_c;

  printf("set cudnn tensor\n");
  checkCUDNN(cudnnSetTensor4dDescriptorEx(ioDesc, dataType, n, c,
      h, w, stride_n, stride_c, stride_h, stride_w));
  checkCUDNN(cudnnSetTensor4dDescriptorEx(bnDesc, dataType, 1, c,
      1, 1, c, 1, 1, 1));

  float alpha, beta;
  alpha = 1.0f;
  beta = 0.0f;
  double epsilon = 1E-5;

  checkCUDNN(cudnnBatchNormalizationForwardInference(cudnnHandle,
                                          mode,
                                          &alpha,
                                          &beta,
                                          ioDesc,
                                          d_input,
                                          ioDesc,
                                          d_output,
                                          bnDesc,
                                          d_scale,
                                          d_bias,
                                          d_estimated_mean,
                                          d_estimated_var,
                                          epsilon));
  checkCUDA(cudaMemcpy(h_output, d_output, (n * c * h * w) * sizeof(float),
      cudaMemcpyDeviceToHost));

  free(h_input);
  free(h_output);
  free(h_scale);
  free(h_bias);
  free(h_estimated_mean);
  free(h_estimated_var);
  checkCUDA(cudaFree(d_input));
  checkCUDA(cudaFree(d_output));
  checkCUDA(cudaFree(d_scale));
  checkCUDA(cudaFree(d_bias));
  checkCUDA(cudaFree(d_estimated_mean));
  checkCUDA(cudaFree(d_estimated_var));

  /* Destroy Handles */
  checkCUDNN( cudnnDestroyTensorDescriptor(ioDesc) );
  checkCUDNN( cudnnDestroyTensorDescriptor(bnDesc) );
  checkCUDNN( cudnnDestroy(cudnnHandle) );
  return 0;
}
```

File: paddle/cuda/src/hl_batch_norm.cu [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                                  const real* bias,
+                                  const real* estimatedMean,
+                                  const real* estimatedVar,
+                                  const double epsilon,
+                                  size_t batchSize,
+                                  size_t channel,
+                                  size_t height,
+                                  size_t width) {
+  dim3 block(256, 1);
+  dim3 grid(1, batchSize);
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_batch_norm.cu]: `gird(batchSize, 1)` is better，Maximum x-dimension is 2^32 - 1, Maximum y dimension is 65536.

File: paddle/cuda/src/hl_batch_norm.cu [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                                   const real* scale,
+                                   const real* bias,
+                                   const real* estimatedMean,
+                                   const real* estimatedVar,
+                                   const double epsilon,
+                                   size_t batchSize,
+                                   size_t channel,
+                                   size_t height,
+                                   size_t width) {
+  const int tid = blockIdx.x * blockDim.x + threadIdx.x;
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_batch_norm.cu]: `blockIdx.x * blockDim.x` can be removed, blockIdx.x is always equal 0.

File: paddle/cuda/src/hl_batch_norm.cu [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  const int num = channel * height * width;
+  const int batch = blockIdx.y;
+  for (int i = tid; i < num; i += blockDim.x) {
+    const int c = (i / (height * width)) % channel;
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_batch_norm.cu]: Can remove `% channel`, i / (height * width) is smaller than the channel.

File: paddle/gserver/layers/CudnnBatchNormLayer.cpp [hunk 4]
```diff
@@ -79,16 +80,32 @@ void CudnnBatchNormLayer::forward(PassType passType) {
...
-                                    ioDesc_,
-                                    output,
-                                    bnParamDesc_,
-                                    gamma,
-                                    beta,
-                                    movingMean,
-                                    movingVar,
-                                    EPS);
+    if (batchSize > 1024) {
+      // there is a bug in cudnn library when the batch size
```
Comment (hedaoyuan) [inline paddle/gserver/layers/CudnnBatchNormLayer.cpp]: Some places say this is a limitation of CUDNN, not bug.

File: paddle/cuda/src/hl_batch_norm.cu (hunk 2 above)
Comment (qingqing01) [inline paddle/cuda/src/hl_batch_norm.cu]: Done.

File: paddle/cuda/src/hl_batch_norm.cu (hunk 3 above)
Comment (qingqing01) [inline paddle/cuda/src/hl_batch_norm.cu]: Done.

File: paddle/cuda/src/hl_batch_norm.cu (hunk 1 above)
Comment (qingqing01) [inline paddle/cuda/src/hl_batch_norm.cu]: Done.

File: paddle/gserver/layers/CudnnBatchNormLayer.cpp (hunk 4 above)
Comment (qingqing01) [inline paddle/gserver/layers/CudnnBatchNormLayer.cpp]: Modify the comments.
